api/src/main/java/org/openmrs/module/patientflags/evaluator/SQLFlagEvaluator.java

```diff
@@ -129,4 +129,73 @@ public FlagValidationResult validate(Flag flag) {
 		// if we've gotten this far, mark the criteria as valid
 		return new FlagValidationResult(true);
 	}
+
+	@Override
+	public String evalMessage(Flag flag, int patientId) {
+		String literal = "\\$\\{\\d{1,2}\\}";
+		String message = flag.getMessage();
+		System.out.println("my message ::::"+message);
+
+		if(!message.matches(".*("+literal+")+.*")){
+			return message;
+		}
+		
+		System.out.println("Replacing values in "+message);
+		
+		Patient p = Context.getPatientService().getPatient(patientId);
+		if(p.isVoided())
+			throw new APIException("VOIDED PATIENT");
+		
+		String criteria = flag.getCriteria();
+		
+		// pull out the "*.patient_id" clause
+		// is this robust enough?
+		Matcher matcher = Pattern.compile("(\\w+\\.patient_id)").matcher(criteria);
+		matcher.find(); // just check for the first occurrence of the pattern... is this enough?
+		String patientIdColumn = matcher.group();
+			
+		// since we are going to append a where/and to the end of this sql statement, we need to trim off trailing ";" and any trailing whitespace
+		matcher = Pattern.compile(";?\\s*$").matcher(criteria);
+		criteria = matcher.replaceFirst(""); // replace first, because there should only be one occurrence
+			
+		// create the criteria for a single patient by appending a "where" or "and" clause
+		String toEval = criteria + (criteria.matches("(?i)(?s).*where.*") ? " and " : " where ") + patientIdColumn + " = "
+			+ p.getPatientId();
+		
+		try {

```

Comment:
Scenario#1:
--------------
We want to highlight a patient who was sick for more than 3 weeks before final illness. The patient could be added to flag list easily by simple query. But if patient passes the flag the duration of illness should also be specified in dashboard.
##### Query:
SELECT p.patient_id, o.value_numeric  
FROM patient p join obs o on o.person_id=p.patient_id  
where o.concept_id=1735 and o.value_numeric > 3;
##### Flag message:
Long final illness duration i.e. ${1} weeks
##### Output:
![pfq1](https://cloud.githubusercontent.com/assets/4829880/6425788/fd79a360-bf5c-11e4-9604-b1b430163ba2.png)

Scenario#2:
--------------
We want to highlight duplicate patients but also want to know the ID which is duplicate.
##### Assumption:
The query may need to be refined more based on different possibilities and also varies across different implementations, but for sake of simplicity we assume that duplicate patient is one sharing same birthdate, birthdate_estimated, deathdate, gender, first_name and identifier_location. 
##### Query:
SELECT 
p.patient_id, duppid.identifier, dupp.person_id, pn.given_name, duppn.given_name 
FROM patient p 
JOIN person pr ON p.patient_id=pr.person_id 
JOIN person_name pn ON p.patient_id=pn.person_id
JOIN patient_identifier pid ON pr.person_id=pid.patient_id 
JOIN person dupp ON pr.person_id <> dupp.person_id 
	AND DATE(dupp.birthdate)=DATE(pr.birthdate) 
	AND dupp.birthdate_estimated=pr.birthdate_estimated 
	AND dupp.dead=pr.dead  
	AND IF(pr.death_date IS NULL, 1, DATE(dupp.death_date)=DATE(pr.death_date)) 
	AND dupp.gender=pr.gender 
JOIN person_name duppn ON duppn.person_id=dupp.person_id AND duppn.given_name=pn.given_name 
LEFT JOIN patient_identifier duppid ON dupp.person_id=duppid.patient_id AND duppid.location_id=pid.location_id;
##### Flag Message:
Possible duplicate of person ID [ ${1} ]
##### Output:
![pfq2](https://cloud.githubusercontent.com/assets/4829880/6425968/5071c182-bf63-11e4-8bf1-1080e568b52c.png)

Similar type of queries could be written for cases when data viewer wants to know hemoglobin reading of patients who lies beyond certain range.